# `NumPy`: Numerical operations in Python

This tutorial uses material also found in the [SciPy 1.0 Nature Methods](https://www.nature.com/articles/s41592-019-0686-2) and [NumPy Array IEEE](10.1109/MCSE.2011.37) article and at [https://numpy.org/](https://numpy.org/).

### Learning outcomes:
 - Understand Python Library `NumPy`
 - Use `numpy arrays`
 - Apply loops and logical operators on `numpy arrays`

Data scientists spend a lot of time – wait for it – working with ***data***! To work with **data** it is critical to organize the data in a way that facilitate the work on the potential analyses we might need to do. So organizing data means guessing what type of work we will want to do with the dataset. And, odd is it may seem, good guessing requires some practice. The data organization process will require: 

* store the data a clear and systematic way
* provide methods to access the data that are simple and straightforward
* be flexible enough so to and allow to modify the format of the data for various needs

NumPy is the fundamental `library` for mathematical operations and computations in Python. 

The NumPy array is a multidimensional array object. A variety of fast operations on arrays are provided by NumPy. These include operations that are mathematical, logical, shape manipulation, sorting, selecting, I/O, discrete Fourier transforms, basic linear algebra, basic statistical operations, random simulation and much more. 

NumPy is the base of scientific computing and data science libraries such as [Pandas](https://pandas.pydata.org/) [scipy.org](https://scipy.org/), and [scikit-learn.org](https://scikit-learn.org/) among many others.

In other (simpler?) words, Numpy arrays are grids or tables for holding, accessing, and manipulating data. They are created and accessed in ways that are very similar to the ways Python `lists` can be accessed.

So what we are going to do is to recall how `lists` work (lists are handy!), then we will graduate to `numpy arrays` and see what they can do. 

### Python lists

We have covered Python `lists` (and other datatypes) in previous tutorials. Python `lists` (a list of things) is build by collecting, ahem, a list of things using `[square brackets]`.

For example:

In [1]:
mylist = ['this', 3, 'list', 4+2j, 6.66]

We can address elements in a list by using indices and the `:` (colon) operator.

In [2]:
mylist[0:3]

['this', 3, 'list']

We can read this as "Give me all the elements in the interval between 0 **inclusive** to 3 **exclusive**."

I know this is weird. But at least for any two indexes `a` and `b`, the number of elements you get back from `mylist[a,b]` is always equal to `b` minus `a`, so I guess that's good!

We can get any consecutive hunk of elements using `:`.

In [3]:
mylist[2:5]

['list', (4+2j), 6.66]

If you omit the indexes, Python will assume you want everything.

In [4]:
mylist[:]

['this', 3, 'list', (4+2j), 6.66]

List can obviously host also homogeneous types of data, such as `int` or `float`:

In [5]:
mylistHomogeneous = [2, 3.14, 10.5, 11.13, 12.7, 4.31]

### Numpy Arrays

Numpy arrays were designed to be lists with superpowers, so everything we learned about  `lists` will apply to `numpy arrays` as well!

A NumPy array is a multidimensional, uniform collection of elements (that is, all elements occupy the same number of bytes in memory). An array is characterized by
 - the type of elements it contains and
 - its shape. 
 
For example, a matrix might be represented as an array of shape M×N that contains numbers, such as floating-point or complex numbers. Unlike matrices, NumPy arrays can have up to 32 dimensions; they might also contain other kinds of elements (or even combinations of elements), such as Booleans or dates. [Ref. Van Der Walt et al. IEEE](10.1109/MCSE.2011.37)

Not to state the obvious, but to use `numpy arrays`, we'll need to `import` the library `numpy`. The standard is to import `numpy` as `np`:

In [6]:
import numpy as np

`arrays` can be made by simply asking for one and filling it out with values, in much the same way we make a `list`.

In [7]:
myarr = np.array([2, 4, 6, 8, 9, 10])

The command `print` can be used also on `NumPy arrays` and it returns the content of the array:

In [8]:
print(myarr)

[ 2  4  6  8  9 10]


By simply returning the array object name (`myarr`) the output is a bit more informative and it returns the type (`array`):

In [9]:
myarr

array([ 2,  4,  6,  8,  9, 10])

From then on, all the indexing we've learned so far applies directly! Square brackets are used for indexing and the same type of addressing can be used as we have learned for `lists`:

In [10]:
myarr[4]

9

In [11]:
myarr[-3:]

array([ 8,  9, 10])

$\color{blue}{\text{Complete the following exercise.}}$

  - Create a `NumPy array` containing both `int`, `float` and `complex` datatypes.

  [Use the cell below to show your code]

In [13]:
myarray = np.array([1, 0.055, 5+3j])

  - Create a `NumPy array` containing `str` as datatypes.

  [Use the cell below to show your code]

In [14]:
myarray2 = np.array(['Jay','Hello'])

### Operations on `numpy arrays`: the difference between `lists` and `arrays`
Indeed, we can make an array directly out of a list.

In [16]:
myArrFromList = np.array(mylistHomogeneous)

In [17]:
myArrFromList

array([ 2.  ,  3.14, 10.5 , 11.13, 12.7 ,  4.31])

And then of course we can index it exactly the same way, so... *Wait, why are we making arrays now? What's the difference?*

One **huge** difference is that if we wanted to do some math with basic Python lists, the fact that they can hold multiple types of data elements does not assure that the mathematical operations will perform.

In [18]:
mylistHomogeneous + 5

TypeError: can only concatenate list (not "int") to list

`numpy` arrays instead contain numerical elements by definition. This definition assures the ability to perform math ith the arrays. So, whereas the addition above did not work when using the `list`, it does work when using the `numpy` array, even though both `list` and `array` contain the same elements!

In [19]:
myArrFromList + 5

array([ 7.  ,  8.14, 15.5 , 16.13, 17.7 ,  9.31])

Now **that** seems like it might be useful!

$\color{blue}{\text{Complete the following exercise.}}$

  - What happens when you add a number to a `NumPy array`? How do the content of the array change?

    - _By using the `NumPy array`, you now can do math with the lists/array in the dataframe._

  - Create a new `array` and multiply the array by a complex number:

  [Use the cell below to show your code]

In [22]:
myarray2 = np.array([2,3,4,5,6])
myarray2 * 2J

array([0. +4.j, 0. +6.j, 0. +8.j, 0.+10.j, 0.+12.j])

_By multiplying with the complex number, they can still generate the answers regard the math values._

### More operations on `arrays`

Two arrays can be added, or subtracted, or multiplied or whatever!

In [23]:
myarr + myArrFromList

array([ 4.  ,  7.14, 16.5 , 19.13, 21.7 , 14.31])

In [24]:
myarr * myArrFromList

array([  4.  ,  12.56,  63.  ,  89.04, 114.3 ,  43.1 ])

In [25]:
myarr / myArrFromList

array([1.        , 1.27388535, 0.57142857, 0.71877808, 0.70866142,
       2.32018561])

We can also **combine** our 2 arrays into a single ***two dimensional (2D) array***.

In [26]:
twoDarr = np.array([myarr, myArrFromList])

In [27]:
twoDarr

array([[ 2.  ,  4.  ,  6.  ,  8.  ,  9.  , 10.  ],
       [ 2.  ,  3.14, 10.5 , 11.13, 12.7 ,  4.31]])

Simple though this may seem, *2D arrays just like this are the bedrock of data analysis!* Arrays of real data are usually larger – sometimes much much larger! – but all the principles are the same and all you as a Data Scientists need to remember is the dimensionality of the data arrays. Python will then compute what you ask for.

But, hold on one second. Remembering the dimensionality of the array is ** *very* ** imoortant. Indeed Python can perform some operations if two arrays do *not* have the same dimensions, but other operations are likely to fail.

For example, imagine two `arrays` with different dimensions:

In [28]:
smallArray = [2, 3, 4]
largeArray = [2, 3, 4, 5, 6, 7]

The two arrays can be added together by using the symbol `+`:

In [29]:
smallArray + largeArray

[2, 3, 4, 2, 3, 4, 5, 6, 7]

Yet, the same two arrays cannot be multiplied:

In [30]:
smallArray * largeArray

TypeError: can't multiply sequence by non-int of type 'list'

This is because Python cannot identify elements to match during the multiplication.

$\color{blue}{\text{Complete the following exercise.}}$

  - What happens when you add two arrays of different dimensions? Say one array with 6 complex numers and one with 4 `float` numbers?

      [Use the cell below to show how to create and add the two arrays]

In [34]:
floatnum = [0.1,0.2,0.3,0.4]
complexnum = [1+1j, 2+1j, 3+1j, 4+1j, 5+1j, 6+1j]
floatnum + complexnum

[0.1, 0.2, 0.3, 0.4, (1+1j), (2+1j), (3+1j), (4+1j), (5+1j), (6+1j)]

_They still combined both dataframe into one._

### Methods of `numpy arrays`

So the shape of the array (the dimensionality) is key, especially if we plan on doing math with the arrays, whihc is the primary goal of the arrays! 

`numpy arrays` are Python objects and as such they have `methods`. A variety of methods exist for the array and `shape` is the one that allow us to retrieve the dimensionality of an array.

In [35]:
twoDarr.shape

(2, 6)

Unlike lists, which are always just lists, arrays can come in any shape. So it's *really* convenient that they can tell us what shape they are straight away.

Indexing into 2D arrays is a straightforward extension of indexing into 1D arrays or lists. We just provide a second index after a `,` (comma). Like this.

In [36]:
twoDarr[1,3]

11.13

The first index refers to the **row index**, and the second to the **column index**. In this case, we're asking for the value in the second row and the fourth column, which is indeed 7 (remember *the first row and column are index=0!*).

We can play all the same games indexing with 2D arrays as we can with 1D arrays, we just have to remember that everything before the comma `,` refers to the *rows* in that it specifies locations along the *vertical dimension*, and everything after the comma `,` refers to the *columns* in that it specifies locations along the *horizontal dimension*.

So this:

In [37]:
twoDarr[:,0:3]

array([[ 2.  ,  4.  ,  6.  ],
       [ 2.  ,  3.14, 10.5 ]])

means "Give me all the rows (the colon `:`) in the first 3 columns (the "`0:3`)."

I told you that the colon all alone by itself would end up being useful!!! In this case for example, by using the `:` you do not need to type many indices (one per row) and you even do not need to remmeber how many rows there are, just use `:` and Python will return all the elements.

A few more examples:

In [38]:
# the last row (regardless of the number of rows, 
# again you do not need to knowhow many rows exist)
twoDarr[-1,:] 

array([ 2.  ,  3.14, 10.5 , 11.13, 12.7 ,  4.31])

In [39]:
twoDarr[:,-2:] # last two columns

array([[ 9.  , 10.  ],
       [12.7 ,  4.31]])

In [40]:
twoDarr[0,::2] # first row, every other column

array([2., 6., 9.])

To get good at this, you don't need natural born talent or anything like that. Like so much in life, the key is *practice, practice, practice*!!! So play around! You can't break your computer or anything!

Another neat trick that arrays can do is *transpose* themselves, flipping the rows for columns.

(Hold your right hand in front of your face so that you're looking at your palm with your fingers pointing towards the left. Now flip your hand so that you're looking at the back of your hand with your fingers pointing up. You just *transposed* your hand such that the first row (your pointer finger) became the first column!)

In [41]:
colarr = twoDarr.T

In [42]:
colarr

array([[ 2.  ,  2.  ],
       [ 4.  ,  3.14],
       [ 6.  , 10.5 ],
       [ 8.  , 11.13],
       [ 9.  , 12.7 ],
       [10.  ,  4.31]])

Why would we want to do that? By convention, *variables* in datasets should correspond to the columns, and *observations* should correspond to the rows. So we have taken data in which this was not so and turned it into an array in which the columns are the first few non-prime numbers and the prime numbers, respectively, and the rows correspond to the instances in order (1st , 2nd, 3rd, ....).

We have just done a little of what is known as **data wrangling**. While not as fun as data visualization, data wrangling is often a big part of any analysis project!

Now that we have the data into shape, we can unleash all the powers of numpy arrays, powers which pandas DataFrames will inherit and build upon!

For example, who's bigger overall, the primes or the non-primes?

In [43]:
colarr.sum(0)

array([39.  , 43.78])

The primes win! 
In `colarr.sum(0)`, the 0 means "the first (vertical) dimension", i.e., sum the values *across the rows* within each column. To sum along the second dimension, we do:

In [44]:
colarr.sum(1)

array([ 4.  ,  7.14, 16.5 , 19.13, 21.7 , 14.31])

So any numpy array knows how to add up the numbers in it by row or by column (see what happens if you leave off the dimension, like this `colarr.sum()`. The list of things that numpy arrays can do themselves is pretty impressive.

Check it out [here](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html).

(or paste this into your browser: https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html)

$\color{blue}{\text{Complete the following exercise.}}$

  - How many methods does a numpy array have? [ _7_ ]
  
  - Create a new 2-dimensional array, and show the use of two methods not used above (`prod` and `round` could be two simple ones, but no pressure):
  
  [Use the cell below to show how to create and add the two arrays]
  
*Hint:* The symbol `?` can be used at the end of a method and that can help understand how to use the method, for example, `myarray.shape?`

In [62]:
np.round(colarr, 2)

array([[ 2.  ,  2.  ],
       [ 4.  ,  3.14],
       [ 6.  , 10.5 ],
       [ 8.  , 11.13],
       [ 9.  , 12.7 ],
       [10.  ,  4.31]])

### `NumPy` methods to create arrays

Often, we want to create a array that we know we're going to put values in later. For example, we might be planning on doing a computation that will result in 3 sets of 7 values, and we want be able to store them directly into an array. We can pre-make an array filled with zeros with `np.zeros(r, c)`.

In [46]:
myzeros = np.zeros((7, 3))

In [47]:
myzeros

array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

Another handy method to create arrays is `ones`

In [48]:
myones = np.ones((3,4))

In [49]:
myones

array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.]])

We will encounter other `NumPy` methods in later tutorials. For the time being one last method that will turn out very handy when modelling data:

In [51]:
myRandomNumArray = np.random.randn(10,1)
print(myRandomNumArray)

[[-1.08500168]
 [ 0.81565774]
 [ 0.26788301]
 [-0.13376991]
 [ 0.25224173]
 [ 0.41907024]
 [ 1.74374693]
 [-0.41162714]
 [-0.17402854]
 [-0.45588912]]


The `numpy` submodule `random` contains a variety of methods to create arrays containing random numbers. Generating random numbers is helpful in many applications, for example, they can be used to create normally distributed noise, or data with normally distributed noise, etc. 

$\color{blue}{\text{Complete the following exercise.}}$
  
  - Create a new 1-dimensional array of uniformly-distributed random number:
  
  [Use the cell below to show your code]


In [94]:
np.random.randn(10)

array([-1.14396057, -0.1236101 , -0.5416752 ,  0.46218554, -0.20007673,
       -0.23184272,  0.45685923,  0.09822733,  0.19791828, -0.62364482])

  
  - Create a new  2-dimensional array of normally-distributed random number:
  
  [Use the cell below to show your code]


In [93]:
np.random.randn(10,2,8)

array([[[-2.47537872e+00, -1.68858560e+00, -5.80699987e-01,
          2.59471729e+00, -6.25964690e-01,  4.52722149e-01,
          1.48801261e+00,  1.48716798e+00],
        [ 8.94978648e-01,  1.10160500e+00,  1.98319463e+00,
         -3.43537086e-01,  1.59741627e+00,  4.21283427e-01,
          1.31408300e+00, -5.59578851e-01]],

       [[ 5.92954913e-01, -4.79485640e-01, -6.82772594e-02,
          3.89615014e-01, -7.16385894e-01, -1.25103925e+00,
          2.42977800e-01, -1.04106222e+00],
        [ 6.30135527e-02,  5.25247330e-01, -5.65709538e-02,
         -1.64163862e+00, -5.31186876e-01, -1.76291319e+00,
         -6.28341165e-01,  1.65505138e+00]],

       [[ 6.39875735e-01,  1.94981001e+00,  1.02949445e+00,
          5.32118356e-01,  1.03391386e+00, -3.90851532e-02,
          5.23657320e-01, -7.06174423e-01],
        [-8.41520709e-01,  2.63828697e-01,  5.32855468e-01,
         -1.34320708e+00,  4.40729450e-01, -1.60891098e-01,
         -1.26272318e+00, -2.00830950e+00]],

       [[-

Let's now create a simple 1-D array and explore what happens when we add a number to the values and what happens when we multiply the numbers in the array:

In [66]:
size  = 20
origArray = np.random.randn(size,1)

Let's look at the values in the array.

In [67]:
print(origArray)

[[-0.07997464]
 [ 0.50422234]
 [-0.65619861]
 [-0.0441136 ]
 [ 0.39986246]
 [ 0.09849765]
 [ 1.11628521]
 [-1.00860885]
 [ 0.10554256]
 [-0.14404121]
 [-1.62817258]
 [ 0.86951907]
 [ 0.69661994]
 [-0.18587627]
 [-0.72068576]
 [ 0.41368476]
 [ 0.005126  ]
 [ 0.05991034]
 [ 1.49530083]
 [-0.22236098]]


There seem to be a variety of numbers, some positive, some negative, as expected because `randn` is supposed to generate numbers centered at 0 (i.e., with mean 0) and standard deviation of 1.

Let's compute the standard deviation and mean of these numbers. Numpy provides to handy methods:

In [68]:
mean = np.mean(origArray)
sd = np.std(origArray)
print(['The mean is:', mean])
print(['the STD is:', sd])

['The mean is:', 0.05372693306108003]
['the STD is:', 0.7073420908608667]


Well, okay, the mean is not quite close to 0, but perhaps close enough? The standard deviation seems pretty close to the expected value of 1.

$\color{blue}{\text{Complete the following exercise.}}$
  
  - Create a new 1-dimensional array of 100 normally-distributed random numbers:
  
  [Use the cell below to show your code]


In [73]:
random_100 = np.random.randn(100,1)
mean_100 = np.mean(random_100)
sd_100 = np.std(random_100)
print("Mean =", mean_100)
print("Std = ", sd_100)

Mean = -0.0798703039702782
Std =  0.9879531574804046


   - What happens to the mean and standard deviation after increasing the size of the array? Are they closer of further from the expected values? Why? 
       - _They are closer to the expected value. The mean us closer to 0. Meahilw, the standard deviation is closer to 1._

What happens if we add 5 to the array?

In [74]:
x  = 5 + origArray

In [75]:
print(x)

[[4.92002536]
 [5.50422234]
 [4.34380139]
 [4.9558864 ]
 [5.39986246]
 [5.09849765]
 [6.11628521]
 [3.99139115]
 [5.10554256]
 [4.85595879]
 [3.37182742]
 [5.86951907]
 [5.69661994]
 [4.81412373]
 [4.27931424]
 [5.41368476]
 [5.005126  ]
 [5.05991034]
 [6.49530083]
 [4.77763902]]


It looks like the values shifted. But how much? It looks like they recentered at 5, the value we added. So we can perhaps assume that now the distribution of numbers is normally distributed but with a mean of 5. The standard deviation has not bee changed. It is still at 1, trust me for the moment and let try multiplying the numbers.

In [76]:
x  = 2 * origArray
print(x)

[[-0.15994927]
 [ 1.00844469]
 [-1.31239722]
 [-0.0882272 ]
 [ 0.79972491]
 [ 0.19699529]
 [ 2.23257041]
 [-2.01721769]
 [ 0.21108512]
 [-0.28808241]
 [-3.25634517]
 [ 1.73903815]
 [ 1.39323988]
 [-0.37175255]
 [-1.44137152]
 [ 0.82736952]
 [ 0.010252  ]
 [ 0.11982069]
 [ 2.99060165]
 [-0.44472197]]


It looks like the values increased. There seem to be a larger variality, more bigger numbers, both negative and positive. So perhaps the STD is not at 1 anymore. Could it be at 2?

$\color{blue}{\text{Complete the following exercise.}}$
  
  - Compute the mean, std and median of `x`:
  
  [Use the cell below to show your code]


In [78]:
mean_x = np.mean(x)
sd_x = np.std(x)
median_x = np.median(x)
print("Mean = ", mean_x)
print("Std =", sd_x)
print("Median =", median_x)

Mean =  0.10745386612216005
Std = 1.4146841817217335
Median = 0.06503634542588894


  
  - What are the mean, std and median of `x`? Why, what is going on here?
      - _The mean is 0.10; the standard deviation is 1.41; the median is 0.06. The mean is larger than the meduian, which means that my data is right skewed_. 

### Summary

So in this tutorial we have shown how to organize and manipulate data using Python `numpy` `arrays`.

So those are the basics of numpy arrays. They:

* store values in rows and columns
* each dimension starts at index zero (like lists)
* can be accessed using
    - square brackets `[]` with row and column indexes separated by a comma
    - integer indexes (including negative "start from the end" indexes)
    - a colon `:` (or two if you want a step value other than 1)
* can have maths done to every element in one go
* can be added, subtracted, etc. from one another
* have superpowers! they can compute stuff along their rows and columns!

The operations that are available for these two data types will be the base for many things that you might need to do as a Data Scientist. 

`Numpy arrays` Can host a variety of data types. Although this might be too much for now, below a table of all the data types an `array` can support:

| Type	| Description |
| --- | --- |
| bool |	Boolean (True or False) stored as a bit (0, 1) |
| inti	| Platform integer (normally either int32 or int64) |
| int8	| Byte (-128 to 127) |
| int16	| Integer (-32768 to 32767) |
| int32	| Integer (-2 ** 31 to 2 ** 31 -1) |
| int64	| Integer (-2 ** 63 to 2 ** 63 -1) |
| uint8	| Unsigned integer (0 to 255) |
| uint16	| Unsigned integer (0 to 65535) |
| uint32	| Unsigned integer (0 to 2 ** 32 – 1) |
| uint64	| Unsigned integer (0 to 2 ** 64 – 1) |
| float16	| Half precision float: sign bit, 5 bits exponent, and 10 bits mantissa |
| float32	| Single precision float: sign bit, 8 bits exponent, and 23 bits mantissa |
| float64 or float	| Double precision float: sign bit, 11 bits exponent, and 52 bits mantissa |
| complex64	| Complex number, represented by two 32-bit floats (real and imaginary components) |
| complex128 or complex	| Complex number, represented by two 64-bit floats (real and imaginary components) |

$\color{blue}{\text{Complete the following exercise.}}$
  
  - Generate an 1-D array of mean = 10 and std = 1.5:
  
  [Use the cell below to show your code]

In [92]:
test_value = np.random.normal(loc = 100, scale = 1.5, size = 1000)
test_mean = np.mean(test_value)
test_sd = np.std(test_value)

print("Mean =", np.round(test_mean,2))
print("Std =", np.round(test_sd,1))

Mean = 100.02
Std = 1.5
